portfolio/src/main/java/com/google/sps/servlets/NicknameServlet.java

```diff
@@ -0,0 +1,105 @@
+// Copyright 2019 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package com.google.sps.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.users.UserService;
+import com.google.appengine.api.users.UserServiceFactory;
+import java.io.IOException;
+import java.io.PrintWriter;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+/**
+   * Class that manages setting and editing users' nicknames.
+   */
+@WebServlet("/nickname")
+public class NicknameServlet extends HttpServlet {
+
+  /* Get the current nickname, if it exists. Returns an empty string otherwise. 
+     This function is used to set the default value in the form for nickname
+     set up, this way is easier for the user to edit it.*/
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    response.setContentType("text/html");
+
+    UserService userService = UserServiceFactory.getUserService();
+
+    String nickname = getUserNickname(userService.getCurrentUser().getUserId());
+    response.getWriter().println(nickname);
+  }
+
+  /**
+   * If the user is not logged in, the form is hidden, so no post request can be
+   * performed, no need for checking if the user is logged.
+   */
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    UserService userService = UserServiceFactory.getUserService();
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    String nickname = request.getParameter("nickname");
+    String id = userService.getCurrentUser().getUserId();
```

Comment:
nit: Note you are assuming here that the user is already authenticated.

If they are not, getCurrentuser() can return nullptr and you will get a null pointer exception, causing your handler to fail and to return a '500 internal server error' to the caller:
https://cloud.google.com/appengine/docs/standard/java/javadoc/com/google/appengine/api/users/UserService#getcurrentuser

I noticed you mentioned in another comment that this can't happen because the form is hidden. While this is true, nothing is preventing a potential abusing user from unhiding the form in chrome developer tools or by doing a manual post request (via a command line tool like curl). Attackers often attempt to send invalid requests to your server in order to 'crash it' in case you're not handling user input correctly (although nothing to bad happens in this particular case).

In general:
- The java code runs on your server. You can trust the server.
- The html/js code runs on your clients browser. You can never trust the client to do what you instructed him to do.

So in this case, I would suggest manually checking that getCurrentUser() does not return null, and returning a 401 http error in case it does.

Reply:
Fixed.